warehouse/accounts/views.py

```diff
@@ -800,9 +800,38 @@ def request_password_reset(request, _form_class=RequestPasswordResetForm):
         if user is None:
             user = user_service.get_user_by_email(form.username_or_email.data)
         if user is not None:
-            email = first_true(
-                user.emails, pred=lambda e: e.email == form.username_or_email.data
+            verified_email = first_true(
+                user.emails,
+                pred=lambda e: e.email == form.username_or_email.data and e.verified,
             )
+
+            if not verified_email:
+                # If the user is found but the email is not verified,
+                # tell the user to contact support instead
+                if unverified_email := first_true(
+                    user.emails,
+                    pred=lambda e: e.email == form.username_or_email.data,
+                ):
+                    user.record_event(
+                        tag=EventTag.Account.PasswordResetAttempt,
+                        request=request,
+                    )
+                    request.log.warning(
+                        "User requested password reset for unverified email",
+                        username=user.username,
+                        email_address=unverified_email.email,
+                    )
+                    request.session.flash(
+                        request._(
+                            "Email address '${email}' is not verified. "
+                            "Contact PyPI support for assistance.",
+                            mapping={"email": unverified_email.email},
+                        ),
+                        queue="error",
+                    )
```

Comment:
I'm also slightly concerned that this will become an oracle. I think instead, if we get a password reset for an unverified email, we should just send that address an email "Password reset requested but this email address is currently unverified" and just display the same ambiguous message here instead.

Reply:
Can you help me understand the threat vector this kind of oracle would provide?
It'll confirm the existence of an email address that isn't able to be used without an account recovery via support.

I'm generally okay to consider another email template to send to unverified email addresses, but that has the downside of us trying to email places that have already had either a hard bounce, expiration, or some other reason that the address is unverified, and could affect our sender scores.

Reply:
We discussed this via chat and confirmed that changing the behavior is desirable.